examples/sensors-viewer/src/components/SensorViewer.jsx

```diff
@@ -0,0 +1,221 @@
+
+class SensorViewer extends Component {
+    state = {
+        visible: false,
+        deviceID: null,
+        data: {},
+        availableSensors: {},
+        sensorValues: {},
+        sensorSamplingRate: {},
+        loading: false,
+        fetched: false,
+    };
+
+    componentDidMount() {
+        if (isMissingCredentials())
+            this.setState({visible: true})
+    }
+
+
+    handleChange = (e) => {
+        e.preventDefault();
+        this.setState({[e.target.name]: e.target.value})
+    };
+
+
+    handleCredentialModal = (visible) => {
+        this.setState({visible});
+        if (!visible) {
+            this.handleSubmit();
+        }
+    };
+
+
+    checkDeviceType = (value) => {
+        const expression = new RegExp(/[a-z]?[A-Z]?[0-9]?-?_?/i);
+        if (value.length === 22)
+            if (expression.test(value))
+                return constant.ID;
+        return constant.ALIAS;
+    };
+
+
+    handleError(err) {
+        if (err.response.status === 403)
+            this.setState({visible: true, loading: false});
+        if (err.response.status === 404) {
+            this.setState({loading: false});
+            window.confirm("Device ID Invalid")
+        }
+    }
+
+
+    handleSubmit = () => {
+        const {deviceID} = this.state;
+        if (deviceID) {
+            const type = this.checkDeviceType(deviceID);
+            this.setState({
+                loading: true,
+                availableSensors: {},
+                sensorValues: {},
+                sensorSamplingRate: {},
+                fetched: false,
+                notFound: false
+            });
+            if (type === constant.ID) {
+                this.setDeviceDataById(deviceID);
+            } else {
+                this.setDeviceDataByAlias(deviceID);
+            }
+        }
+    };
+
+
+    setDeviceDataById = (id, params = {}) => {
+        getDeviceDataById(id, params)
+            .then(res => {
+                    this.setState({
+                        data: res.data,
+                        loading: false,
+                        fetched: true,
+                    });
+                    getInterfaceById(id, res.interfaces[res.availableIndex])
+                        .then(response => {
+                            this.setState({availableSensors: response});
+                        });
+                    getInterfaceById(id, res.interfaces[res.valueIndex])
+                        .then(response => {
+                            this.setState({sensorValues: response})
+                        });
+                    getInterfaceById(id, res.interfaces[res.samplingRateIndex])
+                        .then(response => {
+                            this.setState({sensorSamplingRate: response})
+                        })
+                }
+            )
+            .catch(err => {
+                this.handleError(err)
+            })
+    };
+
+    setDeviceDataByAlias = (alias, params = {}) => {
```

Comment:
setDeviceDataById and setDeviceDataByAlias do very similar thing, but they start from a different state.
It would be better to have a separate function for the common calculations.

Reply:
Both are calling different functional internal "getInterfaceById" and "getInterfaceByAlisas" so we can't make a common function for common code of both functions.